cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/deploy/InMemoryProgramRunDispatcher.java

```diff
@@ -0,0 +1,456 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.io.File;
+import java.io.IOException;
+import java.net.InetAddress;
+import java.nio.file.FileAlreadyExistsException;
+import java.nio.file.Files;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import javax.annotation.Nullable;
+
+/**
+ * In-memory Implementation for {@link ProgramRunDispatcher} which runs the program in the same
+ * service from which it's invoked.
+ */
+public class InMemoryProgramRunDispatcher implements ProgramRunDispatcher {
+
+  private static final String CLUSTER_SCOPE = "cluster";
+  private static final String APPLICATION_SCOPE = "app";
+  private static final Logger LOG = LoggerFactory.getLogger(InMemoryProgramRunDispatcher.class);
+
+  private final CConfiguration cConf;
+  private final Impersonator impersonator;
+  private final ArtifactRepository artifactRepository;
+  private final ProgramRunnerFactory programRunnerFactory;
+  private final ConfiguratorFactory configuratorFactory;
+  private ProgramRunnerFactory remoteProgramRunnerFactory;
+  private String hostname;
+  private boolean isDistributed;
+
+  @Inject
+  public InMemoryProgramRunDispatcher(CConfiguration cConf,
+      ProgramRunnerFactory programRunnerFactory,
+      ConfiguratorFactory configuratorFactory, Impersonator impersonator,
+      @Named(AppFabricServiceRuntimeModule.NOAUTH_ARTIFACT_REPO) ArtifactRepository artifactRepository) {
+    this.cConf = cConf;
+    this.programRunnerFactory = programRunnerFactory;
+    this.configuratorFactory = configuratorFactory;
+    this.impersonator = impersonator;
+    this.artifactRepository = artifactRepository;
+  }
+
+  /**
+   * Optional guice injection for the {@link ProgramRunnerFactory} used for remote execution. It is
+   * optional because in unit-test we don't have need for that.
+   */
+  @Inject(optional = true)
+  public void setRemoteProgramRunnerFactory(
+      @Constants.AppFabric.RemoteExecution ProgramRunnerFactory runnerFactory) {
+    this.remoteProgramRunnerFactory = runnerFactory;
+  }
+
+  @Inject(optional = true)
+  public void setHostname(@Named(Constants.Service.MASTER_SERVICES_BIND_ADDRESS) InetAddress host) {
+    if (Objects.nonNull(host)) {
+      this.hostname = host.getCanonicalHostName();
+    }
+  }
+
+  @Override
+  public ProgramController dispatchProgram(ProgramRunInfo programRunInfo) throws Exception {
+    RunId runId = programRunInfo.getRunId();
+    LOG.debug("Start pipeline launch for Run ID: {}", runId);
```

Comment:
`Preparing to dispatch program run: {}`
Its possible to launch various program types including workers, pipelines

Reply:
Changed.